From ae06b743d39449f7eb0c65a5c9dad1a6802809d8 Mon Sep 17 00:00:00 2001
From: Gustavo Duarte <gduarte@redhat.com>
Date: Wed, 23 Jul 2014 20:54:34 -0400
Subject: [ethernet] cxgb4: allow large buffer size to have page size

Message-id: <1406148874-3009-1-git-send-email-gduarte@redhat.com>
Patchwork-id: 86364
O-Subject: [RHEL7.1 PATCH BZ 1078977] cxgb4: allow large buffer size to have page size
Bugzilla: 1078977
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1078977
Upstream Status: 940d9d34a5467c2e2574866eb009d4cb61e27299
Build Info: https://brewweb.devel.redhat.com/taskinfo?taskID=7727551
Tested: IBM tested this fix on top of 7.0 and verified this fixes the issue.
Sanity checks on top of 7.1 (131) works fine.

commit 940d9d34a5467c2e2574866eb009d4cb61e27299
Author: Thadeu Lima de Souza Cascardo <cascardo@linux.vnet.ibm.com>
Date:   Mon Dec 23 15:34:29 2013 -0200

    cxgb4: allow large buffer size to have page size

    Since commit 52367a763d8046190754ab43743e42638564a2d1
    ("cxgb4/cxgb4vf: Code cleanup to enable T4 Configuration File support"),
    we have failures like this during cxgb4 probe:

    cxgb4 0000:01:00.4: bad SGE FL page buffer sizes [65536, 65536]
    cxgb4: probe of 0000:01:00.4 failed with error -22

    This happens whenever software parameters are used, without a
    configuration file. That happens when the hardware was already
    initialized (after kexec, or after csiostor is loaded).

    It happens that these values are acceptable, rendering fl_pg_order equal
    to 0, which is the case of a hard init when the page size is equal or
    larger than 65536.

    Accepting fl_large_pg equal to fl_small_pg solves the issue, and
    shouldn't cause any trouble besides a possible performance reduction
    when smaller pages are used. And that can be fixed by a configuration
    file.

    Signed-off-by: Thadeu Lima de Souza Cascardo <cascardo@linux.vnet.ibm.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/chelsio/cxgb4/sge.c b/drivers/net/ethernet/chelsio/cxgb4/sge.c
index cc380c3..cc3511a 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/sge.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/sge.c
@@ -2581,7 +2581,7 @@ static int t4_sge_init_soft(struct adapter *adap)
  #undef READ_FL_BUF
 
  if (fl_small_pg != PAGE_SIZE ||
-     (fl_large_pg != 0 && (fl_large_pg <= fl_small_pg ||
+     (fl_large_pg != 0 && (fl_large_pg < fl_small_pg ||
       (fl_large_pg & (fl_large_pg-1)) != 0))) {
   dev_err(adap->pdev_dev, "bad SGE FL page buffer sizes [%d, %d]\n",
    fl_small_pg, fl_large_pg);
-- 
1.7.1